src/grpc/infer_handler.h

```diff
@@ -912,8 +979,24 @@ class InferHandlerState {
 
     // True if there is an ongoing write to the grpc stream
     std::atomic<bool> ongoing_write_;
+
+    // Tracks whether the async notification has been delivered by
+    // completion queue.
+    bool received_notification_;
```

Comment:
Should this be atomic? I think this is concurrently accessed by response callback via `state->IsGrpcContextCancelled()`

Reply:
The response callback only reads the variable. And it is not unsafe to try to read it prematurely before the cancellation notification is registered. The state transition will be ignored.